test/unit/lib/utils/fs/safeMoveFile.test.js

```diff
@@ -0,0 +1,160 @@
+    "Error: EXDEV: cross-device link not permitted, rename '/foo/bar' -> '/bar/baz'"
+  );
+  // This is the important property that safeMoveFile looks for
+  // to fallback to copy then rename behaviour
+  fakeCrossDeviceError.code = 'EXDEV';
+
+  // These other properties aren't used but it doesn't hurt to have an accurate error object
+  fakeCrossDeviceError.errno = -18;
+  fakeCrossDeviceError.syscall = 'rename';
+  fakeCrossDeviceError.path = '/foo/bar';
+  fakeCrossDeviceError.dest = '/bar/baz';
+  return fakeCrossDeviceError;
+};
+
+/**
+ * The name of the file that will be moved
+ */
+const artefactName = 'foo-bar.zip';
+
+// Generate temporary directories once per test run
+// They will be cleaned up when the tests are done
+
+/**
+ * The folder that the test file will be moved to
+ */
+const destinationPath = getTmpDirPath();
+
+/**
+ * The folder that the test file will come from
+ */
+const sourcePath = getTmpDirPath();
+
+/**
+ * The full path to destination where the file is being moved to
+ */
+const destinationFile = join(destinationPath, artefactName);
+
+/**
+ * The full path to the file that is being moved
+ */
+const sourceFile = join(sourcePath, artefactName);
+
+describe('#safeMoveFile()', () => {
+  let renameStub;
+  let copyFileStub;
+  let unlinkStub;
+
+  beforeEach(async () => {
+    renameStub = sinon.stub(fsp, 'rename');
+    copyFileStub = sinon.stub(fsp, 'copyFile');
+    unlinkStub = sinon.stub(fsp, 'unlink');
+    // Allow the fs calls to act as normal until we want to force the rename to fail
+    renameStub.callThrough();
+    copyFileStub.callThrough();
+    unlinkStub.callThrough();
+
+    // Ensure the cache directory exists so that the rename doesn't fail
+    fse.ensureDirSync(sourcePath);
+    fse.ensureDirSync(destinationPath);
+
+    // Write a test file to rename
+    await fse.writeFile(sourceFile, 'source data');
+  });
+
+  afterEach(() => {
+    // Clean up test directories after each test so they don't interfere with each other
+    fse.removeSync(destinationPath);
+    fse.removeSync(sourcePath);
+    // Reset stubbed methods
+    fsp.rename.restore();
+    fsp.copyFile.restore();
+    fsp.unlink.restore();
+  });
+
+  /**
+   * Run the test suite with the provided post assertion function.
+   * The post assertion function will be called after each test scenario is executed
+   * so that the test scenarios can be reused.
+   *
+   * @param {*} postAssertion the function that is called after every test scenario
+   */
+  const runTestScenariosWithPostAssertion = (postAssertion) => {
+    describe('when the file does not exist by the time the zip is generated', () => {
+      it('should write a file', async () => {
+        await safeMoveFile(sourceFile, destinationFile);
+
+        const exists = fse.existsSync(destinationFile);
+        expect(exists).to.be.true;
+        postAssertion();
+      });
+    });
+
+    describe('when the file already exists by the time the zip is generated', () => {
```

Comment:
This util is file type agnostic. So I guess better title name would be simply: `when file at target path already exists`

Reply:
Oh yeah, there would be a lot of these issues because I cut and paste the tests from the generateZip tests. Fixed!